src/data/matrix/kronecker_product_bialgebra.lean

```diff
@@ -0,0 +1,138 @@
+/-
+Copyright (c) 2021 Filippo A. E. Nuccio. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Filippo A. E. Nuccio
+-/
+
+import data.matrix.reindex
+import linear_algebra.matrix
+
+/-!
+# Kronecker product of matrices, see https://en.wikipedia.org/wiki/Kronecker_product
+Two main definitions:
+* Given a commutative semiring α and three α algebras R S β such that β is both a R- and a S-algebra
+  and there are two instances of scalar towers `is_scalar_tower α R β` and `is_scalar_tower α S β`,
+  we define `kronecker_biprod` as
+  the bilinear Kronecker product
+  kronecker_biprod : matrix (l n R) →ₗ[α] (matrix m p S) →ₗ[α] (matrix (l × m) (n × p) (β).
+* In the special case when R=α=S=β, we define kronecker_prod, denoted by ⊗ₖ as the R-linear map
+  ⊗ₖ  : matrix (l n R) →ₗ[R] (matrix m p R) →ₗ[R] (matrix (l × m) (n × p) R).
+
+For both products, we prove that it is associative (in theorems `kronecker_biprod_assoc` and
+`kronecker_prod_assoc`, respectively) as well as the so-called `mixed-product property (in theorems
+`kronecker_biprod_mul` and `kronecker_prod_mul`, respectively).
+-/
+
+namespace matrix_bialgebra
+
+open algebra matrix function
+open_locale matrix big_operators
+
+variables (α : Type*) [comm_semiring α]
+variables (R : Type*) [comm_semiring R]
+variables (S : Type*) [comm_semiring S]
+variables (β : Type*) [comm_semiring β]
+variables [algebra α R] [algebra α S] [algebra α β] [algebra R β] [algebra S β]
+variables {l m n p l' m' n' p' : Type*}
+variables [fintype l] [fintype m] [fintype n] [fintype p]
+variables [fintype l'] [fintype m'] [fintype n'] [fintype p']
+
+def kronecker_biprod [is_scalar_tower α R β] [is_scalar_tower α S β] :
+  (matrix l m R) →ₗ[α] (matrix n p S) →ₗ[α] matrix (l × n) (m × p) β :=
+{ to_fun :=
+  begin
+    intro A,
+    use λ B, λ i j, (algebra_map R β (A i.1 j.1)) * (algebra_map S β (B i.2 j.2)),
```

Comment:
I think it might be easier to start with something like
```lean
def kronecker_map (f : α → β → γ) (A : matrix l m α) (B : matrix n p β) :
  matrix (l × n) (m × p) γ 
| i j := f (A i.1 j.1) (B i.2 j.2)


def kronecker_map_add_hom [add_monoid α] [add_monoid β] [add_comm_monoid γ] (f : α →+ β →+ γ) :
  matrix l m α →+ matrix n p β →+ matrix (l × n) (m × p) γ :=
{ to_fun := λ A,
  { to_fun := λ B, kronecker_map (λ a b, f a b) A B,
    map_zero' := by { ext, apply add_monoid_hom.map_zero },
    map_add' := λ B₁ B₂, by { ext, apply add_monoid_hom.map_add } },
  map_zero' := by { ext, dsimp [kronecker_map], rw [add_monoid_hom.map_zero, add_monoid_hom.zero_apply],  },
  map_add' := λ A₁ A₂, by { ext, dsimp [kronecker_map], rw [add_monoid_hom.map_add, add_monoid_hom.add_apply] } }

-- and again for linear maps
```
and then you don't have to go anywhere near algebra maps until right at the end.